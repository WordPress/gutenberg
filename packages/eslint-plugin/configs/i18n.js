module.exports = {
	plugins: [ '@wordpress' ],
	rules: {
		'@wordpress/valid-sprintf': 'error',
		'@wordpress/i18n-translator-comments': 'error',
		'@wordpress/i18n-text-domain': 'error',
		'@wordpress/i18n-no-collapsible-whitespace': 'error',
		'@wordpress/i18n-no-placeholders-only': 'error',
		'@wordpress/i18n-no-variables': 'error',
		'@wordpress/i18n-ellipsis': 'error',
		'@wordpress/i18n-no-flanking-whitespace': 'error',
		'@wordpress/i18n-hyphenated-range': 'error',
	},
};
