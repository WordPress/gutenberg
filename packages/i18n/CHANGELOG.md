## 2.0.0 (Unreleased)

### Breaking Change

- Change how required built-ins are polyfilled with Babel 7 ([#9171](https://github.com/WordPress/gutenberg/pull/9171)).  If you're using an environment that has limited or no support for ES2015+ such as lower versions of IE then using [core-js](https://github.com/zloirock/core-js) or [@babel/polyfill](https://babeljs.io/docs/en/next/babel-polyfill) will add support for these methods.

## 1.2.0 (2018-07-12)

### New Feature

- Updated build to work with Babel 7 ([#7832](https://github.com/WordPress/gutenberg/pull/7832))

### Internal

- Moved `@WordPress/packages` repository to `@WordPress/gutenberg` ([#7805](https://github.com/WordPress/gutenberg/pull/7805))

## 1.1.1 (2018-05-18)

### Polish

- Fix: Standardized `package.json` format  ([#119](https://github.com/WordPress/packages/pull/119))
