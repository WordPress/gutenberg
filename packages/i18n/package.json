{
	"name": "@wordpress/i18n",
	"version": "3.5.0",
	"description": "WordPress internationalization (i18n) library.",
	"author": "The WordPress Contributors",
	"license": "GPL-2.0-or-later",
	"keywords": [
		"wordpress",
		"i18n"
	],
	"homepage": "https://github.com/WordPress/gutenberg/tree/master/packages/i18n/README.md",
	"repository": {
		"type": "git",
		"url": "https://github.com/WordPress/gutenberg.git",
		"directory": "packages/i18n"
	},
	"bugs": {
		"url": "https://github.com/WordPress/gutenberg/issues"
	},
	"main": "build/index.js",
	"module": "build-module/index.js",
	"react-native": "src/index",
	"bin": {
		"pot-to-php": "./tools/pot-to-php.js"
	},
	"dependencies": {
		"@babel/runtime": "^7.4.4",
		"gettext-parser": "^1.3.1",
		"lodash": "^4.17.14",
		"memize": "^1.0.5",
		"sprintf-js": "^1.1.1",
		"tannin": "^1.1.0"
	},
	"publishConfig": {
		"access": "public"
	}
}
