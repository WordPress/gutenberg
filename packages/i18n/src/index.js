/**
 * External dependencies
 */
import Tannin from 'tannin';
import memoize from 'memize';
import sprintfjs from 'sprintf-js';

/**
 * Default locale data to use for Tannin domain when not otherwise provided.
 * Assumes an English plural forms expression.
 *
 * @type {Object}
 */
const DEFAULT_LOCALE_DATA = {
	'': {
		plural_forms: ( n ) => n === 1 ? 0 : 1,
	},
};

/**
 * Log to console, once per message; or more precisely, per referentially equal
 * argument set. Because Jed throws errors, we log these to the console instead
 * to avoid crashing the application.
 *
 * @param {...*} args Arguments to pass to `console.error`
 */
const logErrorOnce = memoize( console.error ); // eslint-disable-line no-console

/**
 * The underlying instance of Tannin to which exported functions interface.
 *
 * @type {Tannin}
 */
const i18n = new Tannin( {} );

/**
 * Merges locale data into the Tannin instance by domain. Accepts data in a
 * Jed-formatted JSON object shape.
 *
 * @see http://messageformat.github.io/Jed/
 *
 * @param {?Object} data   Locale data configuration.
 * @param {?string} domain Domain for which configuration applies.
 */
export function setLocaleData( data, domain = 'default' ) {
	i18n.data[ domain ] = {
		...DEFAULT_LOCALE_DATA,
		...i18n.data[ domain ],
		...data,
	};

	// Populate default domain configuration (supported locale date which omits
	// a plural forms expression).
	i18n.data[ domain ][ '' ] = {
		...DEFAULT_LOCALE_DATA[ '' ],
		...i18n.data[ domain ][ '' ],
	};
}

/**
 * Wrapper for Tannin's `dcnpgettext`. Populates default locale data if not
 * otherwise previously assigned.
 *
 * @param {?string} domain  Domain to retrieve the translated text.
 * @param {?string} context Context information for the translators.
 * @param {string}  single  Text to translate if non-plural. Used as fallback
 *                          return value on a caught error.
 * @param {?string} plural  The text to be used if the number is plural.
 * @param {?number} number  The number to compare against to use either the
 *                          singular or plural form.
 *
 * @return {string} The translated string.
 */
function dcnpgettext( domain = 'default', context, single, plural, number ) {
	if ( ! i18n.data[ domain ] ) {
		setLocaleData( undefined, domain );
	}

	return i18n.dcnpgettext( domain, context, single, plural, number );
}

/**
 * Retrieve the translation of text.
 *
 * @see https://developer.wordpress.org/reference/functions/__/
 *
 * @param {string}  text   Text to translate.
 * @param {?string} domain Domain to retrieve the translated text.
 *
 * @return {string} Translated text.
 */
export function __( text, domain ) {
	return dcnpgettext( domain, undefined, text );
}

/**
 * Retrieve translated string with gettext context.
 *
 * @see https://developer.wordpress.org/reference/functions/_x/
 *
 * @param {string}  text    Text to translate.
 * @param {string}  context Context information for the translators.
 * @param {?string} domain  Domain to retrieve the translated text.
 *
 * @return {string} Translated context string without pipe.
 */
export function _x( text, context, domain ) {
	return dcnpgettext( domain, context, text );
}

/**
 * Translates and retrieves the singular or plural form based on the supplied
 * number.
 *
 * @see https://developer.wordpress.org/reference/functions/_n/
 *
 * @param {string}  single The text to be used if the number is singular.
 * @param {string}  plural The text to be used if the number is plural.
 * @param {number}  number The number to compare against to use either the
 *                         singular or plural form.
 * @param {?string} domain Domain to retrieve the translated text.
 *
 * @return {string} The translated singular or plural form.
 */
export function _n( single, plural, number, domain ) {
	return dcnpgettext( domain, undefined, single, plural, number );
}

/**
 * Translates and retrieves the singular or plural form based on the supplied
 * number, with gettext context.
 *
 * @see https://developer.wordpress.org/reference/functions/_nx/
 *
 * @param {string}  single  The text to be used if the number is singular.
 * @param {string}  plural  The text to be used if the number is plural.
 * @param {number}  number  The number to compare against to use either the
 *                          singular or plural form.
 * @param {string}  context Context information for the translators.
 * @param {?string} domain  Domain to retrieve the translated text.
 *
 * @return {string} The translated singular or plural form.
 */
export function _nx( single, plural, number, context, domain ) {
	return dcnpgettext( domain, context, single, plural, number );
}

/**
 * Returns a formatted string. If an error occurs in applying the format, the
 * original format string is returned.
 *
 * @param {string}   format  The format of the string to generate.
 * @param {...string} args Arguments to apply to the format.
 *
 * @see http://www.diveintojavascript.com/projects/javascript-sprintf
 *
 * @return {string} The formatted string.
 */
export function sprintf( format, ...args ) {
	try {
		return sprintfjs.sprintf( format, ...args );
	} catch ( error ) {
		logErrorOnce( 'sprintf error: \n\n' + error.toString() );

		return format;
	}
}
