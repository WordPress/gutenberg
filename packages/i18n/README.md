# Internationalization (i18n)

Internationalization utilities for client-side localization.

https://codex.wordpress.org/I18n_for_WordPress_Developers

## Installation

Install the module:

```bash
npm install @wordpress/i18n --save
```

_This package assumes that your code will run in an **ES2015+** environment. If you're using an environment that has limited or no support for ES2015+ such as lower versions of IE then using [core-js](https://github.com/zloirock/core-js) or [@babel/polyfill](https://babeljs.io/docs/en/next/babel-polyfill) will add support for these methods. Learn more about it in [Babel docs](https://babeljs.io/docs/en/next/caveats)._

## Usage

```js
import { sprintf, _n } from '@wordpress/i18n';

sprintf( _n( '%d hat', '%d hats', 4, 'text-domain' ), 4 );
// 4 hats
```

Note that you will not need to specify [domain](https://codex.wordpress.org/I18n_for_WordPress_Developers#Text_Domains) for the strings.

## Build

You can use the [WordPress i18n babel plugin](../babel-plugin-makepot/README.md) to generate a `.pot` file containing all your localized strings.

The package also includes a `pot-to-php` script used to generate a php files containing the messages listed in a `.pot` file. This is useful to trick WordPress.org translation strings discovery since at the moment, WordPress.org is not capable of parsing strings directly from JavaScript files.

```sh
npx pot-to-php languages/myplugin.pot languages/myplugin-translations.php text-domain
```

## API

`__( text: string, domain: string ): string`

Retrieve the translation of text.

See: https://developer.wordpress.org/reference/functions/__/

`_x( text: string, context: string, domain: string ): string`

Retrieve translated string with gettext context.

See: https://developer.wordpress.org/reference/functions/_x/

`_n( single: string, plural: string, number: Number, domain: string ): string`

Translates and retrieves the singular or plural form based on the supplied number.

See: https://developer.wordpress.org/reference/functions/_n/

`_nx( single: string, plural: string, number: Number, context: string, domain: string ): string`

Translates and retrieves the singular or plural form based on the supplied number, with gettext context.

See: https://developer.wordpress.org/reference/functions/_nx/

`sprintf( format: string, ...args: mixed[] ): string`

Returns a formatted string.

See: http://www.diveintojavascript.com/projects/javascript-sprintf

`setLocaleData( data: Object, domain: string )`

Creates a new Jed instance with specified locale data configuration.

`getI18n(): Jed`

Returns the current Jed instance, initializing with a default configuration if not already assigned.

See: http://messageformat.github.io/Jed/

<br/><br/><p align="center"><img src="https://s.w.org/style/images/codeispoetry.png?1" alt="Code is Poetry." /></p>
