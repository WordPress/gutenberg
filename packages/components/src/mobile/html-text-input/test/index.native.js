/**
 * External dependencies
 */
import { shallow } from 'enzyme';

/**
 * WordPress dependencies
 */
import { __ } from '@wordpress/i18n';

/**
 * Internal dependencies
 */
import { HTMLTextInput } from '..';

// Utility to find a TextInput in a ShallowWrapper
const findTextInputInWrapper = ( wrapper, matchingProps ) => {
	return wrapper.dive().findWhere( ( node ) => {
		return node.name() === 'TextInput' && node.is( matchingProps );
	} ).first();
};

// Finds the Content TextInput in our HTMLInputView
const findContentTextInput = ( wrapper ) => {
	const placeholder = __( 'Start writingâ€¦' );
	const matchingProps = { multiline: true, placeholder };
	return findTextInputInWrapper( wrapper, matchingProps );
};

// Finds the Title TextInput in our HTMLInputView
const findTitleTextInput = ( wrapper ) => {
	const placeholder = __( 'Add title' );
	return findTextInputInWrapper( wrapper, { placeholder } );
};

describe( 'HTMLTextInput', () => {
	it( 'HTMLTextInput renders', () => {
		const wrapper = shallow(
			<HTMLTextInput />
		);
		expect( wrapper ).toBeTruthy();
	} );

	it( 'HTMLTextInput updates store and state on HTML text change', () => {
		const onChange = jest.fn();

		const wrapper = shallow(
			<HTMLTextInput
				onChange={ onChange }
			/>
		);

		expect( wrapper.instance().state.isDirty ).toBeFalsy();

		// Simulate user typing text
		const htmlTextInput = findContentTextInput( wrapper );
		htmlTextInput.simulate( 'changeText', 'text' );

		//Check if the onChange is called and the state is updated
		expect( onChange ).toHaveBeenCalledTimes( 1 );
		expect( onChange ).toHaveBeenCalledWith( 'text' );

		expect( wrapper.instance().state.isDirty ).toBeTruthy();
		expect( wrapper.instance().state.value ).toEqual( 'text' );
	} );

	it( 'HTMLTextInput persists changes in HTML text input on blur', () => {
		const onPersist = jest.fn();

		const wrapper = shallow(
			<HTMLTextInput
				onPersist={ onPersist }
				onChange={ jest.fn() }
			/>
		);

		// Simulate user typing text
		const htmlTextInput = findContentTextInput( wrapper );
		htmlTextInput.simulate( 'changeText', 'text' );

		//Simulate blur event
		htmlTextInput.simulate( 'blur' );

		//Normally prop.value is updated with the help of withSelect
		//But we don't have it in tests so we just simulate it
		wrapper.setProps( { value: 'text' } );

		//Check if the onPersist is called and the state is updated
		expect( onPersist ).toHaveBeenCalledTimes( 1 );
		expect( onPersist ).toHaveBeenCalledWith( 'text' );

		expect( wrapper.instance().state.isDirty ).toBeFalsy();

		//We expect state.value is getting propagated from prop.value
		expect( wrapper.instance().state.value ).toEqual( 'text' );
	} );

	it( 'HTMLTextInput propagates title changes to store', () => {
		const editTitle = jest.fn();

		const wrapper = shallow(
			<HTMLTextInput
				editTitle={ editTitle }
			/>
		);

		// Simulate user typing text
		const textInput = findTitleTextInput( wrapper );
		textInput.simulate( 'changeText', 'text' );

		//Check if the setTitleAction is called
		expect( editTitle ).toHaveBeenCalledTimes( 1 );
		expect( editTitle ).toHaveBeenCalledWith( 'text' );
	} );
} );

