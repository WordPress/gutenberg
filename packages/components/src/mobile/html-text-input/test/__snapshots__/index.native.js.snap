// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HTMLTextInput HTMLTextInput renders and matches snapshot 1`] = `
<View
  onLayout={[Function]}
>
  <RCTScrollView>
    <View>
      <TextInput
        accessibilityLabel="html-view-title"
        autoCorrect={false}
        numberOfLines={1}
        placeholder="Add title"
        placeholderTextColor="white"
        style={
          [
            {
              "color": "white",
            },
            undefined,
          ]
        }
        textAlignVertical="center"
      />
      <TextInput
        accessibilityLabel="html-view-content"
        autoCorrect={false}
        multiline={true}
        onBlur={[Function]}
        onChangeText={[Function]}
        placeholder="Start writingâ€¦"
        placeholderTextColor="white"
        rejectResponderTermination={false}
        scrollEnabled={false}
        style={
          [
            {
              "color": "white",
            },
            undefined,
          ]
        }
        textAlignVertical="top"
      />
    </View>
  </RCTScrollView>
</View>
`;
